clients/vault/src/oracle/storage/impls.rs

```diff
@@ -230,6 +230,82 @@ impl ScpArchiveStorage {
 	}
 }
 
+impl TransactionsArchiveStorage {
+	pub async fn get_transactions_archive(slot_index: i32) -> Result<XdrArchive<TransactionHistoryEntry>, Error> {
+		let (url, file_name) = Self::get_url_and_file_name(slot_index);
+		//try to find xdr.gz file and decode. if error then download archive from horizon archive
+		// node and save
+		println!("{url}");
+		println!("{file_name}");
+		let result = Self::try_gz_decode_archive_file(&file_name);
+
+		if result.is_err() {
+			let result = Self::download_file_and_save(&url, &file_name).await;
+			if result.is_ok() {
+				let data = Self::try_gz_decode_archive_file(&file_name)?;
+				return Ok(Self::decode_xdr(data))
+			}
+		}
+		let data = result.unwrap();
+		Ok(Self::decode_xdr(data))
+	}
+
+	fn decode_xdr(xdr_data: Vec<u8>) -> XdrArchive<TransactionHistoryEntry> {
+		XdrArchive::<TransactionHistoryEntry>::from_xdr(xdr_data).unwrap()
+	}
+
+	async fn download_file_and_save(url: &str, file_name: &str) -> Result<(), Error> {
+		let response = reqwest::get(url).await.unwrap();
+		let content = response.bytes().await.unwrap();
+
+		let mut file = match File::create(&file_name) {
+			Err(why) => panic!("couldn't create {}", why),
+			Ok(file) => file,
+		};
+		file.write_all(content.as_bytes_ref())?;
+		Ok(())
+	}
+
+	fn try_gz_decode_archive_file(path: &str) -> Result<Vec<u8>, Error> {
+		use flate2::bufread::GzDecoder;
+		use std::io::{self, BufReader, Read};
+		let bytes = Self::read_file_xdr(path)?;
+		let mut gz = GzDecoder::new(&bytes[..]);
+		let mut bytes: Vec<u8> = vec![];
+		gz.read_to_end(&mut bytes)?;
+		Ok(bytes)
+	}
+
+	fn get_url_and_file_name(slot_index: i32) -> (String, String) {
+		let slot_index = Self::find_last_slot_index_in_batch(slot_index);
+		let hex_string = format!("0{:x}", slot_index);
+		let file_name = format!("{hex_string}.xdr");
+		let base_url = crate::oracle::constants::stellar_history_base_url_transactions;
+		let url = format!(
+			"{base_url}{}/{}/{}/transactions-{file_name}.gz",
+			&hex_string[..2],
+			&hex_string[2..4],
+			&hex_string[4..6]
+		);
+		(url, format!("txs-{file_name}"))
+	}
+
+	fn find_last_slot_index_in_batch(slot_index: i32) -> i32 {
+		let rest = (slot_index + 1) % 64;
```

Comment:
let's try to avoid magic numbers, I assume this 64 is the batch size? Anyway it'd be nice to have a constant. specially if it's also used somewhere else.

Reply:
@gonzamontiel 
You are rigth! Thanks for advise!